splash-dao-offchain/src/state_projection.rs

```diff
@@ -1,22 +1,457 @@
+use std::sync::Arc;
+
+use async_std::task::spawn_blocking;
 use bloom_offchain::execution_engine::bundled::Bundled;
-use spectrum_offchain::data::event::{AnyMod, Predicted, Traced};
-use spectrum_offchain::data::{EntitySnapshot, Identifier};
+use rocksdb::{Direction, IteratorMode, ReadOptions};
+use serde::de::DeserializeOwned;
+use serde::{Deserialize, Serialize};
+use spectrum_offchain::data::event::{AnyMod, Confirmed, Predicted, Traced};
+use spectrum_offchain::data::{EntitySnapshot, HasIdentifier};
+use spectrum_offchain::rocks::RocksConfig;
 
 /// Projection of [T] state relative to the ledger.
 #[async_trait::async_trait]
 pub trait StateProjectionRead<T, B>
 where
-    T: EntitySnapshot,
+    T: EntitySnapshot + HasIdentifier,
+    T::Id: Send + Serialize,
 {
-    async fn read<I>(&self, id: I) -> Option<AnyMod<Bundled<T, B>>>
-    where
-        I: Identifier<For = T> + Send;
```

Comment:
Why HasIdentifier is better than Identifier<For = T>?

Reply:
Resolved since `HasIdentifier` has been deleted.